File: common/mrtjp/projectred/core/utils/MathLib.java
package mrtjp.projectred.core.utils;

import java.security.InvalidParameterException;

import net.minecraft.entity.Entity;
import net.minecraft.util.MathHelper;

import org.lwjgl.util.vector.Matrix4f;
import org.lwjgl.util.vector.Vector3f;

import codechicken.lib.vec.Vector3;

public class MathLib
{
    public static Matrix4f createEntityRotateMatrix(Entity entity)
    {
        double yaw = Math.toRadians(entity.rotationYaw - 180);
        double pitch = Math.toRadians(entity.rotationPitch);

        Matrix4f initial = new Matrix4f();
        initial.rotate((float) pitch, new Vector3f(1, 0, 0));
        initial.rotate((float) yaw, new Vector3f(0, 1, 0));
        return initial;
    }

    public static Vector3 getLook(Entity source, float f)
    {
        if (f == 1.0F)
        {
            float var2 = MathHelper.cos(-source.rotationYaw * 0.01745329F - 3.141593F);
            float var3 = MathHelper.sin(-source.rotationYaw * 0.01745329F - 3.141593F);
            float var4 = -MathHelper.cos(-source.rotationPitch * 0.01745329F);
            float var5 = MathHelper.sin(-source.rotationPitch * 0.01745329F);
            return new Vector3(var3 * var4, var5, var2 * var4);
        }

        float var2 = source.prevRotationPitch + (source.rotationPitch - source.prevRotationPitch) * f;
        float var3 = source.prevRotationYaw + (source.rotationYaw - source.prevRotationYaw) * f;
        float var4 = MathHelper.cos(-var3 * 0.01745329F - 3.141593F);
        float var5 = MathHelper.sin(-var3 * 0.01745329F - 3.141593F);
        float var6 = -MathHelper.cos(-var2 * 0.01745329F);
        float var7 = MathHelper.sin(-var2 * 0.01745329F);
        return new Vector3(var5 * var6, var7, var4 * var6);
    }

    public static Vector3 bezier(Vector3 s, Vector3 c1, Vector3 c2, Vector3 e, float t)
    {
        if ((t < 0.0F) || (t > 1.0F))
            return s;

        float one_minus_t = 1.0F - t;

        Vector3 retValue = new Vector3(0.0D, 0.0D, 0.0D);
        Vector3[] terms = new Vector3[4];
        terms[0] = calcNewVector(one_minus_t * one_minus_t * one_minus_t, s);
        terms[1] = calcNewVector(3.0F * one_minus_t * one_minus_t * t, c1);
        terms[2] = calcNewVector(3.0F * one_minus_t * t * t, c2);
        terms[3] = calcNewVector(t * t * t, e);

        for (int i = 0; i < 4; i++)
            retValue.add(terms[i]);

        return retValue;
    }

    private static Vector3 calcNewVector(float scaler, Vector3 base)
    {
        Vector3 retValue = new Vector3(base.x, base.y, base.z);
        retValue.multiply(scaler);
        return retValue;
    }
}

File: src/core/scala/mrtjp/core/math/MathLib.scala
/*
 * Copyright (c) 2014.
 * Created by MrTJP.
 * All rights reserved.
 */
package mrtjp.core.math

import java.util.Random

import codechicken.lib.vec.Vector3
import net.minecraft.util.math.BlockPos

object MathLib
{
    def clamp(min:Float, max:Float, v:Float) = Math.min(max, Math.max(min, v))

    def normal(bc:BlockPos, dir:Int):(Int, Int) = normal(bc.getX, bc.getY, bc.getZ, dir)
    def normal(x:Int, y:Int, z:Int, dir:Int):(Int, Int) = dir match
    {
        case 0 => (x, z)
        case 1 => (x, z)
        case 2 => (x, y)
        case 3 => (x, y)
        case 4 => (y, z)
        case 5 => (y, z)
    }

    def basis(bc:BlockPos, dir:Int):Int = basis(bc.getX, bc.getY, bc.getZ, dir)
    def basis(x:Int, y:Int, z:Int, dir:Int):Int = dir match
    {
        case 0 => y
        case 1 => y
        case 2 => z
        case 3 => z
        case 4 => x
        case 5 => x
    }

    def shift(dir:Int) = if ((dir&1) == 1) 1 else -1

    def splitLine(xs:Seq[Int], shift:Int) =
    {
        if (xs.isEmpty) Seq()
        else
        {
            var start = 0
            val ret = for
            {
                (x, i) <- xs.zipWithIndex
                if i > 0
                if x != xs(i-1)+shift
            } yield
            {
                val size = i-start
                start = i
                (xs(i-1), size)
            }
            ret :+ ((xs.last, xs.length-start))
        }
    }

    def rhrAxis(dir:Int, normal:(Int, Int), basis:Int):BlockPos = dir match
    {
        case 0 => new BlockPos(normal._1, basis, normal._2)
        case 1 => new BlockPos(normal._1, basis, normal._2)
        case 2 => new BlockPos(normal._1, normal._2, basis)
        case 3 => new BlockPos(normal._1, normal._2, basis)
        case 4 => new BlockPos(basis, normal._1, normal._2)
        case 5 => new BlockPos(basis, normal._1, normal._2)
    }

    def bezier(s:Vector3, c1:Vector3, c2:Vector3, e:Vector3, t:Float):Vector3 =
    {
        if ((t < 0.0F) || (t > 1.0F)) return s
        val one_minus_t = 1.0F-t
        val retValue = new Vector3(0.0D, 0.0D, 0.0D)
        val terms = new Array[Vector3](4)

        def calcNewVector(scaler:Float, base:Vector3) = base.copy.multiply(scaler)

        terms(0) = calcNewVector(one_minus_t*one_minus_t*one_minus_t, s)
        terms(1) = calcNewVector(3.0F*one_minus_t*one_minus_t*t, c1)
        terms(2) = calcNewVector(3.0F*one_minus_t*t*t, c2)
        terms(3) = calcNewVector(t*t*t, e)

        for (i <- 0 until 4) retValue.add(terms(i))
        retValue
    }

    private val random = new Random
    def randomFromIntRange(az:Range, rand:Random = random) = az(rand.nextInt(az.size))

    def leastSignificant(mask:Int) =
    {
        var bit = 0
        var m = mask
        while ((m&1) == 0 && m != 0){ bit += 1; m <<= 1 }
        bit
    }

    def mostSignificant(mask:Int):Int =
    {
        if (mask == 0) return 0
        31-Integer.numberOfLeadingZeros(mask)
    }

    def weightedRandom[T](xs:Iterable[(T, Int)], rand:Random = random):T =
    {
        if (xs.size == 1) return xs.head._1
        var weight = rand.nextInt(xs.map(_._2).sum)
        xs.find(x => {weight -= x._2; weight < 0}).get._1
    }
}